src/mpi/init/init_async.c

```diff
@@ -59,22 +74,163 @@ static void progress_fn(void *data)
     return;
 }
 
+#if defined(MPL_HAVE_MISC_GETNPROCS)
+MPL_STATIC_INLINE_PREFIX int MPIDI_parse_progress_thread_affinity(int *thread_affinity,
+                                                                  int async_threads_per_node)
+{
+    int th_idx, read_count = 0, mpi_errno = MPI_SUCCESS;
+    char *affinity_copy = NULL;
+    const char *affinity_to_parse = MPIR_CVAR_CH4_PROGRESS_THREAD_AFFINITY;
+    char *proc_id_str, *tmp;
+    size_t proc_count;
+
+    if (!affinity_to_parse || strlen(affinity_to_parse) == 0) {
+        printf("ERROR: unexpected affinity string\n");
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+
+    /* create copy of original buffer because it will be modified in strsep */
+    affinity_copy = MPL_strdup(affinity_to_parse);
+    MPIR_Assert(affinity_copy);
+    tmp = affinity_copy;
+
+    /* apply auto affinity */
+    if (!strcmp(affinity_copy, "auto")) {
+        /* generate default affinity */
+        proc_count = MPL_get_nprocs();
+        for (th_idx = 0; th_idx < async_threads_per_node; th_idx++) {
+            if (th_idx < proc_count)
+                thread_affinity[th_idx] = proc_count - th_idx - 1;
+            else
+                thread_affinity[th_idx] = thread_affinity[th_idx % proc_count];
+        }
+        goto fn_exit;
+    }
+
+    /* apply explicit affinity to the provided logical cpu cores */
+    for (th_idx = 0; th_idx < async_threads_per_node; th_idx++) {
+        proc_id_str = MPL_strsep(&tmp, ",");
+        if (proc_id_str != NULL) {
+            if ((strlen(proc_id_str) > 0 && !isdigit(proc_id_str[0])) || atoi(proc_id_str) < 0) {
+                printf
+                    ("ERROR: unexpected proc_id %s, affinity string %s\n",
+                     proc_id_str, affinity_to_parse);
```

Comment:
Don't do `printf`, just use error code and long error message. Just use a standard error message that prints the entire (failed) affinity string.

Reply:
Should we print the same high level error message with the entire failed affinity string for the different places during parsing or also keep the details that are currently there in the different cases?

Reply:
Same error message should be fine.

Reply:
Done